examples/fabric-admin/rpc/RpcClient.cpp

```diff
@@ -0,0 +1,168 @@
+pw::rpc::Channel channels[] = { pw::rpc::Channel::Create<1>(&hdlc_channel_output) };
+
+// Initialize the RPC client with the channels.
+pw::rpc::Client client(channels);
+
+// Generated clients are namespaced with their proto library.
+using FabricBridgeClient = chip::rpc::pw_rpc::nanopb::FabricBridge::Client;
+
+// RPC channel ID on which to make client calls. RPC calls cannot be made on
+// channel 0 (Channel::kUnassignedChannelId).
+constexpr uint32_t kDefaultChannelId = 1;
+
+// Function to process incoming packets
+void ProcessPackets()
+{
+    std::array<std::byte, kMaxTransmissionUnit> inputBuf;
+    pw::hdlc::Decoder decoder(inputBuf);
+
+    while (true)
+    {
+        std::array<std::byte, kMaxTransmissionUnit> data;
+        auto ret = rpcSocketStream.Read(data);
+        if (!ret.ok())
+        {
+            if (ret.status() == pw::Status::OutOfRange())
+            {
+                // Handle remote disconnect
+                rpcSocketStream.Close();
+                return;
+            }
+            continue;
+        }
+
+        for (std::byte byte : ret.value())
+        {
+            auto result = decoder.Process(byte);
+            if (!result.ok())
+            {
+                // Wait for more bytes that form a complete packet
+                continue;
+            }
+            pw::hdlc::Frame & frame = result.value();
+            if (frame.address() != pw::hdlc::kDefaultRpcAddress)
+            {
+                // Wrong address; ignore the packet
+                continue;
+            }
+
+            client.ProcessPacket(frame.data()).IgnoreError();
+        }
+    }
+}
+
+template <typename CallType>
+CHIP_ERROR WaitForResponse(CallType & call)
+{
+    if (!call.active())
+    {
+        return CHIP_ERROR_INTERNAL;
+    }
+
+    // Wait for the response or timeout
+    uint32_t elapsedTimeMs     = 0;
+    const uint32_t sleepTimeMs = 100;
+
+    while (call.active() && elapsedTimeMs < kRpcTimeoutMs)
+    {
+        usleep(sleepTimeMs * 1000);
+        elapsedTimeMs += sleepTimeMs;
+    }
+
+    if (elapsedTimeMs >= kRpcTimeoutMs)
+    {
+        fprintf(stderr, "RPC Response timed out!");
+        return CHIP_ERROR_TIMEOUT;
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+void AddDeviceResponse(const pw_protobuf_Empty & response, pw::Status status)
+{
+    if (status.ok())
+    {
+        printf("RPC call succeeded\n");
+    }
+    else
+    {
+        printf("RPC call failed with status: %d\n", status.code());
+    }
+}
+
+} // namespace
+
+CHIP_ERROR InitRpcClient(uint16_t rpcServerPort)
+{
+    if (rpcSocketStream.Connect(rpcServerAddress, rpcServerPort) != PW_STATUS_OK)
+    {
+        ChipLogError(NotSpecified, "Failed to connect the Fabric-Admin");
+        return CHIP_ERROR_NOT_CONNECTED;
+    }
+
+    ChipLogProgress(NotSpecified, "Connectted to the Fabric-Admin\n");
+
+    // Start a thread to process incoming packets
+    std::thread packet_processor(ProcessPackets);
```

Comment:
Event handling using the existing scheme could/should be done. Right now, the RPCs end-up doing calls to Matter stack context without external synchronization.

Reply:
When using pw_rpc in Pigweed, it is generally recommended to dedicate a separate thread to handle the ProcessPacket function. This ensures that the main thread or other threads are not blocked by potentially time-consuming packet processing, enhancing the overall performance and responsiveness of your application.

In our cases, the main thread of Fabric Bridge is pending on CHIP event, and the main thread of Fabric Admin is pending on command input, if we call ProcessPacket from those thread, we never receive the RPC callbacks. 

Pigweed allows calling RPC functions from threads other than the one handling ProcessPacket, the pw_rpc framework is designed to handle concurrency, but you must ensure that the underlying transport and any shared resources are managed safely. Specifically, the Send function of a ChannelOutput must not call any RPC APIs or interact with other pw_rpc endpoints to avoid deadlock situations. ​